javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -658,36 +622,45 @@ public Double getNext() {
         } else if (step == 0) {
             throw new IllegalArgumentException("step cannot be 0");
         } else if (from == toInclusive) {
-            return Iterator.of(from);
-        } else if (step * (from - toInclusive) > 0) {
-            return Iterator.empty();
+            return of(from);
+        } else if (Math.signum(step) == Math.signum(from - toInclusive)) {
+            return empty();
         } else {
-            return new AbstractIterator<Double>() {
-
-                double prev = Double.NaN;
-                double curr = from;
-                boolean hasNext = true;
+            if (step > 0) {
+                return new AbstractIterator<Double>() {
+                    final double end = Math.nextUp(toInclusive);
+                    double i = from;
+
+                    @Override
+                    public boolean hasNext() {
+                        return i < end;
+                    }
 
-                @Override
-                public boolean hasNext() {
-                    return hasNext;
-                }
+                    @Override
+                    public Double getNext() {
+                        final double next = i;
+                        i = Math.max(i + step, Math.nextUp(i));
```

Comment:
Scala seems to use BigDecimal under the hood - the problem is with the CPU floating point representation, I think we should try to make it as intuitive as possible.

@danieldietrich, @ruslansennov: What would a user find more intuitive:
* a loop that stops before the end because consecutive numbers have a greater distance than the step size
* if it stepped bigger in the last few steps
* if it's slower (uses `BigDecimal`) and has a few consecutive values that are equal?